# SVD da matriz de classificação de filmes  

A matriz \(M\) contém as avaliações (1–5 estrelas) dos **4 usuários** para **5 filmes**  
(valores 0 = filme ainda não avaliado).

$$
M=\begin{bmatrix}
5 & 4 & 0 & 0 & 3\\
3 & 0 & 4 & 2 & 1\\
4 & 3 & 5 & 5 & 0\\
0 & 2 & 1 & 3 & 4
\end{bmatrix}
$$
Fatoraremos $$(M = U\,\Sigma\,V^{\mathsf T})$$ usando a Decomposição em Valores Singulares (SVD).


In [ ]:
# Cálculo do SVD (Singular Value Decomposition) da matriz
import numpy as np

M = np.array([
    [5, 4, 0, 0, 3],
    [3, 0, 4, 2, 1],
    [4, 3, 5, 5, 0],
    [0, 2, 1, 3, 4]
], dtype=float)

# SVD compacta  (k = min(n_users, n_filmes) = 4)
U, S, Vt = np.linalg.svd(M, full_matrices=False)

# Exibição organizada
np.set_printoptions(precision=3, suppress=True)
print("U (4×4):\n", U, "\n")
print("Σ (4×4 diagonal):\n", np.diag(S), "\n")
print("Vᵀ (4×5):\n", Vt)


U (4×4):
 [[ 0.452  0.808 -0.378 -0.01 ]
 [ 0.42  -0.312 -0.141 -0.841]
 [ 0.715 -0.439 -0.097  0.536]
 [ 0.329  0.24   0.91  -0.077]] 

Σ (4×4 diagonal):
 [[11.483  0.     0.     0.   ]
 [ 0.     5.625  0.     0.   ]
 [ 0.     0.     4.094  0.   ]
 [ 0.     0.     0.     2.177]] 

Vᵀ (4×5):
 [[ 0.556  0.402  0.486  0.47   0.269]
 [ 0.24   0.426 -0.569 -0.372  0.546]
 [-0.66   0.004 -0.034  0.48   0.578]
 [-0.198  0.648 -0.349  0.352 -0.543]]


### Valores obtidos (arredondados a 3 casas)

$$U \approx
\begin{bmatrix}
\;\;0.452 & 0.808 & -0.378 & -0.010\\
\;\;0.420 & -0.312 & -0.141 & -0.841\\
\;\;0.715 & -0.439 & -0.097 & \;\;0.536\\
\;\;0.329 & 0.240 & \;\;0.910 & -0.077
\end{bmatrix}$$


$$\Sigma \approx \operatorname{diag}\bigl(11.483,\; 5.625,\; 4.094,\; 2.177\bigr)$$


$$V^{\mathsf T} \approx
\begin{bmatrix}
\;\;0.556 & 0.402 & 0.486 & 0.470 & 0.269\\
\;\;0.240 & 0.426 & -0.569 & -0.372 & 0.546\\
-0.660 & 0.004 & -0.034 & 0.480 & 0.578\\
-0.198 & 0.648 & -0.349 & 0.352 & -0.543
\end{bmatrix}
$$

> Os quatro valores singulares (11.483, 5.625, 4.094, 2.177) indicam que os **dois primeiros fatores latentes** concentram a maior parte da variância; portanto, uma versão truncada com \(k=2\) ou \(k=3\) pode ser usada para prever notas faltantes e gerar recomendações mais robustas.
